### PR TITLE
feat: Onboard American Community Survey dataset

### PR DESCRIPTION
**Pull request by dipannitab2392**
## Description
Pipeline List : 
- census_bureau_acs/cbsa_2019_5yr
- census_bureau_acs/congressionaldistrict_2019_1yr
- census_bureau_acs/congressionaldistrict_2019_5yr
- census_bureau_acs/county_2019_1yr
- census_bureau_acs/county_2019_5yr
- census_bureau_acs/place_2019_1yr
- census_bureau_acs/place_2019_5yr
- census_bureau_acs/cbsa_2019_1yr
- census_bureau_acs/puma_2019_1yr
- census_bureau_acs/puma_2019_5yr
- census_bureau_acs/schooldistrictelementary_2019_1yr
- census_bureau_acs/schooldistrictelementary_2019_5yr
- census_bureau_acs/schooldistrictsecondary_2019_1yr
- census_bureau_acs/schooldistrictsecondary_2019_5yr
- census_bureau_acs/schooldistrictunified_2019_1yr
- census_bureau_acs/schooldistrictunified_2019_5yr
- census_bureau_acs/state_2019_1yr
- census_bureau_acs/state_2019_5yr
- census_bureau_acs/zcta_2019_5yr

## Checklist

Note: Delete items below that aren't applicable to your pull request.

- [x] Please merge this PR for me once it is approved.
- [x] If this PR adds or edits a dataset or pipeline, it was reviewed and approved by the Google Cloud Public Datasets team beforehand.
- [x] If this PR adds or edits a dataset or pipeline, I put all my code inside  `datasets/census_bureau_acs` and nothing outside of that directory.
- [x] This PR is appropriately labeled.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -5,6 +5,7 @@ name = "pypi"
 
 [packages]
 beautifulsoup4 = "==4.9.3"
+"google.cloud" = "*"
```
Comment (adlersantos) [inline Pipfile:8]: Adding this will apply to all pipelines, perhaps only add it under your image's `requirements.txt` file

Comment (dipannitab2392) [reply Pipfile:8]: @adlersantos - already Pipfile has "google.cloud" = "*"

Comment (adlersantos) [reply Pipfile:8]: Let's remove this because this one's included in Cloud Composer

**Review by adlersantos — CHANGES_REQUESTED** [review]
Set the schedule to `@once`

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+
+    logging.info(
+        f"ACS {pipeline_name} process started at "
+        + str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+    )
+
+    logging.info("Creating 'files' folder")
+    pathlib.Path("./files").mkdir(parents=True, exist_ok=True)
+
+    group_id = {
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:53]: Let's put this in a `.json` file in the same directory as this file, and then load it

```
# File is group_ids.json
group_id = json.load("group_ids.json")
```

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+        "C24060004": "sales_office_employed",
+        "C24060002": "management_business_sci_arts_employed",
+        "B23006001": "pop_25_64",
+        "B23006023": "bachelors_degree_or_higher_25_64",
+        "B11001007": "nonfamily_households",
+        "B11001002": "family_households",
+        "B25035001": "median_year_structure_built",
+    }
+
+    state_code = {
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:306]: We can the same here (separate file loaded)

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+def create_geo_id(df: pd.DataFrame, concat_col: str) -> pd.DataFrame:
+    df["geo_id"] = ""
+    for col in concat_col:
+        df["geo_id"] = df["geo_id"] + df[col]
+    return df
+
+
+def change_length(val: str, length: int) -> str:
+    if len(str(val)) < int(length):
+        return ("0" * (int(length) - len(str(val)))) + str(val)
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:512]:
```suggestion
    if len(str(val)) < length:
        return ("0" * (length - len(str(val)))) + str(val)
```

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+    df = df.replace(to_replace={"KPI_Name": group_id})
+
+    logging.info("Renaming headers...")
+    rename_headers(df, rename_mappings)
+
+    logging.info("Creating column geo_id...")
+    if geography == "censustract" or geography == "blockgroup":
+        df["tract"] = df["tract"].apply(change_length, args=("6"))
+        df["state"] = df["state"].apply(change_length, args=("2"))
+        df["county"] = df["county"].apply(change_length, args=("3"))
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:385]:
```suggestion
        df["tract"] = df["tract"].apply(change_length, args=(6))
        df["state"] = df["state"].apply(change_length, args=(2))
        df["county"] = df["county"].apply(change_length, args=(3))
```
and perhaps use the function name `pad_zeroes_to_the_right` or something in that regard, to make it more descriptive

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+    list_temp = []
+    for key in group_id:
+        logging.info(f"reading data from API for KPI {key}...")
+        # replece = {
+        #     "+year_report+": year_report,
+        #     "+key[0:-3]+": key[0:-3],
+        #     "+key[-3:]+": key[-3:],
+        #     "+api_naming_convention+": api_naming_convention,
+        # }
+        # source_url_new = string_replace(source_url, replece)
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:433]: Let's remove this

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+        for sc in state_code:
+            logging.info(f"reading data from API for KPI {key}...")
+            logging.info(f"reading data from API for KPI {sc}...")
+            # replece = {
+            #     "+year_report+": y ear_report,
+            #     "+key[0:-3]+": key[0:-3],
+            #     "+key[-3:]+": key[-3:],
+            #     "+api_naming_convention+": api_naming_convention,
+            #     "+sc+": sc,
+            # }
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:474]: ^

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+            str1 = source_url.replace("+year_report+", year_report)
+            str2 = str1.replace("+key[0:-3]+", key[0:-3])
+            str3 = str2.replace("+key[-3:]+", key[-3:])
+            str4 = str3.replace("+api_naming_convention+", api_naming_convention)
+            source_url_new = str4.replace("+sc+", sc)
+            # source_url = string_replace(source_url, replece)
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:480]: ^

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+    blob = bucket.blob(gcs_path)
+    blob.upload_from_filename(file_path)
+
+
+if __name__ == "__main__":
+    logging.getLogger().setLevel(logging.INFO)
+
+    main(
+        source_url=os.environ["SOURCE_URL"],
+        # source_file=pathlib.Path(os.environ["SOURCE_FILE"]).expanduser(),
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:538]: ^

File: datasets/census_bureau_acs/cbsa_2019_1yr/pipeline.yaml [hunk 10]
```diff
@@ -0,0 +1,825 @@
...
+              nodeSelectorTerms:
+                - matchExpressions:
+                    - key: cloud.google.com/gke-nodepool
+                      operator: In
+                      values:
+                        - "pool-e2-standard-4"
+        image_pull_policy: "Always"
+        image: "{{ var.json.census_bureau_acs.container_registry.run_csv_transform_kub }}"
+        env_vars:
+          SOURCE_URL: "https://api.census.gov/data/2019/acs/acs+year_report+?get=NAME,+key[0:-3]+_+key[-3:]+E&for=+api_naming_convention+:*&key=550e53635053be51754b09b5e9f5009c94aa0586"
```
Comment (adlersantos) [inline datasets/census_bureau_acs/cbsa_2019_1yr/pipeline.yaml:55]: Let's replace the `key` ranges with more descriptive, parameterized names

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+        logging.info(f"reading data from API for KPI {key}...")
+        # replece = {
+        #     "+year_report+": year_report,
+        #     "+key[0:-3]+": key[0:-3],
+        #     "+key[-3:]+": key[-3:],
+        #     "+api_naming_convention+": api_naming_convention,
+        # }
+        # source_url_new = string_replace(source_url, replece)
+        str1 = source_url.replace("+year_report+", year_report)
+        str2 = str1.replace("+key[0:-3]+", key[0:-3])
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:435]: Use descriptive names for these character ranges

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+        str3 = str2.replace("+key[-3:]+", key[-3:])
+        source_url_new = str3.replace("+api_naming_convention+", api_naming_convention)
+        try:
+            r = requests.get(source_url_new, stream=True)
+            logging.info(f"Source url : {source_url_new}")
+            logging.info(f"status code : {r.status_code}")
+            if r.status_code == 200:
+                text = r.json()
+                frame = pd.DataFrame(text)
+                frame = frame.iloc[1:, :]
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:445]: Extract function to what describes what's happening, e.g. `load_nested_json_into_df_without_headers`

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+                    text = r.json()
+                    frame = pd.DataFrame(text)
+                    frame = frame.iloc[1:, :]
+                    frame["KPI_Name"] = key
+                    list_temp.append(frame)
+            except OSError as e:
+                logging.info(f"error : {e}")
+            except TypeError as error:
+                logging.inf(error)
+                time.sleep(5)
```
Comment (adlersantos) [inline datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py:495]: Let's remove this. Or the whole except block

**Review by adlersantos — CHANGES_REQUESTED** [review]

File: Pipfile (hunk 1 above)
Comment (dipannitab2392) [reply Pipfile]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 2 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: created separate json file

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 3 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: created separate json file

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 4 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 5 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 6 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 7 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

File: datasets/census_bureau_acs/cbsa_2019_1yr/pipeline.yaml (hunk 10 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/cbsa_2019_1yr/pipeline.yaml]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 8 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 9 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 8 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 11 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 12 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: done

File: datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py (hunk 13 above)
Comment (dipannitab2392) [reply datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py]: removed

**Review by adlersantos — APPROVED** [review]

**Pull request merged**